Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,239 @@
+public actor ServiceRunner: Sendable {
+    /// The internal state of the ``ServiceRunner``.
+    private enum State {
+        /// The initial state of the runner.
+        case initial
+        /// The state once ``ServiceRunner/run()`` has been called.
+        case running
+        /// The state once ``ServiceRunner/run()`` has finished.
+        case finished
+    }
+
+    /// The services to run.
+    private let services: [any Service]
+    /// The runner's configuration.
+    private let configuration: ServiceRunnerConfiguration
+    /// The logger.
+    private let logger: Logger
+
+    /// The current state of the runner.
+    private var state: State = .initial
+
+    /// Initializes a new ``ServiceRunner``.
+    ///
+    /// - Parameters:
+    ///   - services: The services to run.
+    ///   - configuration: The runner's configuration.
+    ///   - logger: The logger.
+    public init(
+        services: [any Service],
+        configuration: ServiceRunnerConfiguration,
+        logger: Logger
+    ) {
+        self.services = services
+        self.configuration = configuration
+        self.logger = logger
+    }
+
+    /// Runs all the services by spinning up a child task per service.
+    /// Furthermore, this method sets up the correct signal handlers
+    /// for graceful shutdown.
+    public func run() async throws {
+        switch self.state {
+        case .initial:
+            self.state = .running
+            try await self._run()
+
+            switch self.state {
+            case .initial, .finished:
+                fatalError("ServiceRunner is in an invalid state \(self.state)")
+
+            case .running:
+                self.state = .finished
+            }
+
+        case .running:
+            throw ServiceRunnerError.alreadyRunning()
+
+        case .finished:
+            throw ServiceRunnerError.alreadyFinished()
+        }
+    }
+
+    private func _run() async throws {
+        self.logger.info(
+            "Starting service lifecycle",
+            metadata: [
+                self.configuration.loggingConfiguration.signalsKey: "\(self.configuration.gracefulShutdownSignals)",
+                self.configuration.loggingConfiguration.servicesKey: "\(self.services)",
+            ]
+        )
+
+        enum ChildTaskResult {
+            case serviceFinished(any Service)
+            case serviceThrew(any Service, any Error)
+            case signalCaught(UnixSignal)
+            case signalSequenceFinished
+        }
+
+        try await withThrowingTaskGroup(of: ChildTaskResult.self) { group in
+            // First we have to register our signals.
+            let unixSignals = await UnixSignalsSequence(trapping: self.configuration.gracefulShutdownSignals)
+
+            group.addTask {
+                for await signal in unixSignals {
+                    return .signalCaught(signal)
+                }
+
+                return .signalSequenceFinished
+            }
```

Comment:
Would it make more sense to check if `unixSignals` is not empty before we start that task? In this case we could remove the `signalSequenceFinished`. and just fatal after the for loop.

Reply:
I am a bit hesitant to make that change. The `signal` sequence might finish if for example the signal handler get overridden.